In [1]:
import pandas as pd
from sqlalchemy import create_engine

# Анализ базы данных

Компания вышла на книжный рынок, приобретя крупный сервис для чтения книг по подписке. Первостепенная задача аналитиков проанализировать базу данных, хранащую информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. Все это необходимо для возможности формировать ценые предложения для нового продукта.

## Коннекция  и выгрузка данных

Создадим соединения с базой данных SQL по средством библиотеки sqlalchemy, а также выгрузим данные для первостепенной оценки.

In [2]:
db_config = {'user':'praktikum_student',
             'pwd':'Sdf4$2;d-d30pp',
             'host':'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port':6432, 
             'db':'data-analyst-final-project-db'
             }

In [3]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                       db_config['pwd'],
                                                       db_config['host'],
                                                       db_config['port'],
                                                       db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
query = '''SELECT *
           FROM books;'''

books = pd.io.sql.read_sql(query, con = engine)
print(f'Количество дубликатов: {books.duplicated().sum()}')
books

Количество дубликатов: 0


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [5]:
books[books['title'] == 'Memoirs of a Geisha']

book_id  author_id                title  num_pages publication_date  \
425      426         39  Memoirs of a Geisha        434       2005-11-15   
426      427         39  Memoirs of a Geisha        503       2005-11-22   

     publisher_id  
425           241  
426           311

In [6]:
query = '''SELECT *
           FROM authors;'''

authors = pd.io.sql.read_sql(query, con = engine)
print(f'Количество дубликатов: {authors.duplicated().sum()}')
authors

Количество дубликатов: 0


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [7]:
query = '''SELECT *
           FROM publishers;'''

publishers = pd.io.sql.read_sql(query, con = engine)
print(f'Количество дубликатов: {publishers.duplicated().sum()}')
publishers

Количество дубликатов: 0


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [8]:
query = '''SELECT *
           FROM ratings;'''

ratings = pd.io.sql.read_sql(query, con = engine)
print(f'Количество дубликатов: {ratings.duplicated().sum()}')
ratings

Количество дубликатов: 0


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [9]:
query = '''SELECT *
           FROM reviews;'''

reviews = pd.io.sql.read_sql(query, con = engine)
print(f'Количество дубликатов: {reviews.duplicated().sum()}')
reviews

Количество дубликатов: 0


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Соединение создано и теперь имеется представление о хранимых в ней данных.

## Создание SQL-запросов

Сформируем все SQL-запросы которые помогут ответить на вопросы из поставленных задач.

Задачи:
 - Посчитайте, сколько книг вышло после 1 января 2000 года;

In [10]:
query_1 = '''
            SELECT 
                COUNT(book_id)
                 
            FROM 
                books
                     
            WHERE
                CAST(publication_date AS timestamp) > '2000-01-01;'
          '''

 - Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [11]:
query_2 = '''


        SELECT
            title,
            reviews_count,
            avg_rating
            
        FROM
            (
            SELECT 
                 ratings.book_id AS id,
                 count(reviews.book_id) as reviews_count,
                 AVG(ratings.rating) as avg_rating
                             
             FROM 
                 ratings
                 LEFT JOIN reviews ON ratings.book_id = reviews.book_id 
                 AND ratings.username = reviews.username
                 
             GROUP BY
                 ratings.book_id
                 
             ORDER BY
                 id ) AS tbl
            INNER JOIN books ON books.book_id = tbl.id
            
        ORDER BY
            reviews_count DESC;
             
          '''

 - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц;

In [12]:
query_3 = '''

                    SELECT 
                        publisher,
                        count(book_id) as books_count
                        
                    FROM 
                        books
                        INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
                 
                    WHERE
                        num_pages > 50
             
                    GROUP BY
                        publisher
                        
                    ORDER BY
                        books_count DESC
                        
                    LIMIT 1;
          '''

 - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [13]:
query_4 = '''

            SELECT
                author,
                AVG(avg_rating)

            FROM
                (
                SELECT 
                    MAX(authors.author) AS author,
                    COUNT(books.book_id) AS rating_count,
                    AVG(ratings.rating) AS avg_rating
                        
                FROM 
                    books
                    INNER JOIN authors ON books.author_id = authors.author_id
                    INNER JOIN ratings ON books.book_id = ratings.book_id
                
                GROUP BY
                    books.book_id
                    
                HAVING
                    COUNT(books.book_id) >= 50
                    
                ORDER BY
                    avg_rating DESC
                ) as tbl
                
            GROUP BY
                author
                
            ORDER BY
                avg DESC
                
            LIMIT 1;
            
          '''

 - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
query_5 = '''
            
            SELECT
                AVG(text_count)
            
            FROM
                (
                SELECT 
                    ratings.username as user,
                    COUNT(text) AS text_count
                        
                FROM 
                    ratings
                    LEFT JOIN reviews ON ratings.book_id = reviews.book_id
                    AND ratings.username = reviews.username
                
                GROUP BY
                    ratings.username
                    
                HAVING
                    COUNT(rating) > 50
                ) as tbl;
                                   
          '''

Запросы сформированы, выведем их результаты и сделаем вывод по кождому из них.

## Вывод результатов запроса

In [15]:
exercize_1 = pd.io.sql.read_sql(query_1, con = engine)
exercize_1

count
0    819

Итак, после 1 января 2020 года вышло 819 книг.

In [16]:
exercize_2 = pd.io.sql.read_sql(query_2, con = engine)
exercize_2

title  reviews_count  \
0                              Twilight (Twilight  #1)              7   
1                                  Water for Elephants              6   
2                                     The Glass Castle              6   
3    Harry Potter and the Prisoner of Azkaban (Harr...              6   
4    The Curious Incident of the Dog in the Night-Time              6   
..                                                 ...            ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel              0   
996                            The Natural Way to Draw              0   
997   The Cat in the Hat and Other Dr. Seuss Favorites              0   
998                          Essential Tales and Poems              0   
999  Disney's Beauty and the Beast (A Little Golden...              0   

     avg_rating  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 3 columns]

Наибольшее число отзывов у книги Twilight(Twilight # 1), правда средний рейтинг не самый лучший.

In [17]:
exercize_3 = pd.io.sql.read_sql(query_3, con = engine)
exercize_3

publisher  books_count
0  Penguin Books           42

Penguin Books - издательство, выпустившее наибольшее число книг (42) толще 50 страниц.

In [18]:
exercize_4 = pd.io.sql.read_sql(query_4, con = engine)
exercize_4

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPre

In [19]:
exercize_5 = pd.io.sql.read_sql(query_5, con = engine)
exercize_5

avg
0  24.333333

Среднее количество обзоров от пользователей, поставивших более 50 оценок - 24.33.

## Вывод

Благодара запоросам из базы данных SQL удалось подробно изучить имеющиеся данные по выпущенным книгам, а также издательствам выпустившим этим книги, отзывах по этим книгам и наиболее востребованным авторам. Теперь сервис для чтения книг может сформировать маркетинговую политику, зная какие книги являются наиболее популярными и читаемыми.